WebExtensionContext::openTabs() should only return tabs that are open

kiaraarose opened the pull request:
#### 37b7a1c15d162636bf8ed3f38269e4fed3370878
<pre>
WebExtensionContext::openTabs() should only return tabs that are open
<a href="https://bugs.webkit.org/show_bug.cgi?id=274486">https://bugs.webkit.org/show_bug.cgi?id=274486</a>
<a href="https://rdar.apple.com/128285195">rdar://128285195</a>

Reviewed by Timothy Hatcher.

Filter WebExtensionContext::openTabs() to only return tabs that are open.
Not having this check is a bug and adding this check should cause bugs like
<a href="https://rdar.apple.com/128285195">rdar://128285195</a> to reproduce less since didOpenTab wouldn&apos;t have been called
for those tabs since they belong to a different profile.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::openTabs const):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::mainWebView() const):
Remove the THROW_UNLESS from here since throwing here doesn&apos;t give
us any indication of where the issue is occurring from the app.
A better place for this error would be in getOrCreateTab.

Canonical link: <a href="https://commits.webkit.org/279202@main">https://commits.webkit.org/279202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b071178d71684ad754b183677b24678e6c646464

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/52798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32135 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3521 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/38831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3244 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/42859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2276 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/54896 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/29856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45576 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23974 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/26946 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1680 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/48821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3046 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57669 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/27938 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29159 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/45770 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30075 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7741 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/28912 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -2007,6 +2007,11 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
 {
     return WTF::compactMap(m_tabMap, [&](auto& entry) -> std::optional<Ref<WebExtensionTab>> {
+        if (!entry.value->isOpen()) {
+            ASSERT_NOT_REACHED("The tab should be open at this point; please file a bug.");
```
This might still fire in normal cases, if an app call things like `didChangeTabProperties:` before `didOpenTab:`. It would not be a bug.

The logging I think we should add is in `WebExtensionContext::getOrCreateTab`. That will be the stack trace the app bug is happening.

If a `WebExtensionTab` is created or gotten that has `webViews()` with `webView.configuration._webExtensionController != extensionContext()->extensionController()->wrapper()` then we will know what code in the app is causing it (instead of finding out later when `webViews()` is used in unrelated code.) That is the better place to have the `THROW_UNLESS` instead of in `webViews()`.

xeenon replied:
Instead of `THROW_UNLESS` we can use `RELEASE_LOG_FAULT` to get a stack trace (simulated crash) without crashing. That will be actionable info if done in `getOrCreateTab()`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3490,6 +3496,17 @@ static inline bool isNotRunningInTestRunner()
     wakeUpBackgroundContentIfNecessary(WTFMove(completionHandler));
 }
 
+void WebExtensionContext::reportWebViewConfigurationErrorIfNeeded(Ref<WebExtensionTab> tab) const
```
`void WebExtensionContext::reportWebViewConfigurationErrorIfNeeded(const WebExtensionTab& tab) const`

kiaraarose committed:
WebExtensionContext::openTabs() should only return tabs that are open
https://bugs.webkit.org/show_bug.cgi?id=274486
rdar://128285195

Reviewed by Timothy Hatcher.

Filter WebExtensionContext::openTabs() to only return tabs that are open.
Not having this check is a bug and adding this check should cause bugs like
rdar://128285195 to reproduce less since didOpenTab wouldn't have been called
for those tabs since they belong to a different profile.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::openTabs const):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::mainWebView() const):
Remove the THROW_UNLESS from here since throwing here doesn't give
us any indication of where the issue is occurring from the app.
A better place for this error would be in getOrCreateTab.

Canonical link: https://commits.webkit.org/279202@main
